libraries/botbuilder-core/src/turnContext.ts

```diff
@@ -483,29 +484,40 @@ export class TurnContext {
      */
     public sendActivities(activities: Partial<Activity>[]): Promise<ResourceResponse[]> {
         let sentNonTraceActivity = false;
-        const ref: Partial<ConversationReference> = TurnContext.getConversationReference(this.activity);
-        const output: Partial<Activity>[] = activities.map((a: Partial<Activity>) => {
-            const o: Partial<Activity> = TurnContext.applyConversationReference({ ...a }, ref);
-            if (!o.type) {
-                o.type = ActivityTypes.Message;
+
+        const ref = TurnContext.getConversationReference(this.activity);
+        const outputActivities = activities.map((activity) => {
+            const outputActivity = TurnContext.applyConversationReference({ ...activity }, ref);
+
+            if (!outputActivity.type) {
+                outputActivity.type = ActivityTypes.Message;
             }
-            if (o.type !== ActivityTypes.Trace) {
+
+            if (outputActivity.type !== ActivityTypes.Trace) {
                 sentNonTraceActivity = true;
             }
-            if (o.id) {
-                delete o.id;
+
+            if (outputActivity.id) {
+                delete outputActivity.id;
             }
-            return o;
+
+            return outputActivity;
         });
 
-        return this.emit(this._onSendActivities, output, () => {
+        return this.emit(this._onSendActivities, outputActivities, async () => {
             if (this.activity.deliveryMode === DeliveryModes.ExpectReplies) {
-                // Append activities to buffer
-                const responses: ResourceResponse[] = [];
-                output.forEach((a) => {
-                        this.bufferedReplyActivities.push(a);
-                        responses.push({ id: undefined });
-                });
+                // Append ouput activities to buffer and mock responses
+                this.bufferedReplyActivities.push(...outputActivities);
+                const responses = outputActivities.map<ResourceResponse>(() => ({ id: undefined }));
+
+                // Ensure the TurnState has the InvokeResponseKey, since this activity
+                // is not being sent through the adapter, where it would be added to TurnState.
+                const invokeResponse = outputActivities.find(
+                    (activity) => activity.type === ActivityTypes.InvokeResponse
+                );
+                if (invokeResponse) {
+                    this.turnState.set(INVOKE_RESPONSE_KEY, invokeResponse);
```

Comment:
@stevengum and/or @EricDahlvang I'm pretty sure this is roughly equivalent to https://github.com/microsoft/botbuilder-dotnet/pull/4845/files#diff-391e4084826390583f2af2b2d27d0d0add457734d88f05e817e826af88a2b41aR328.